### PR TITLE
Kernel: Implement multiple Region mprotect

### LINKED ISSUE #177: Kernel: Implement mprotect()
We should support altering the protection of VM ranges after they've been created. This will get a bit complicated, since protection bits are currently per-region, so `mprotect()` will need to know how to merge and split regions.

See: http://pubs.opengroup.org/onlinepubs/9699919799/functions/mprotect.html

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
We were able to do this with munmap already, and for mprotect its just copy and paste from there

Fixes #177 and maybe the rest of #964

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
oops messed up the history

File: Kernel/Syscalls/mmap.cpp
Comment (ADKaster) [inline Kernel/Syscalls/mmap.cpp:406]: I think you're missing a return 0 here?

Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:406]: ohh ehm yes, I will fix tomorow

Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:406]: should be fixed

**Comment by Hendiadyoin1** [comment]
some things before merge:  
1. Do we need to check first, or are we allowed to fail half way through?
2. Do we need to assert full coverage, or is it allowed, that there is a not mapped region in between?
3. the issue talked about merging, what is meant with that, are we supposed to merge these multiple regions if we can or do we need to assert if we can etc...

File: Kernel/Syscalls/mmap.cpp
Comment (ADKaster) [inline Kernel/Syscalls/mmap.cpp:384]: I'm not convinced this is correct.. some MAP_FIXED tests are in order I think.  I Would want to verify that all the edge cases work out, such as:

1) middle of one region into the middle of another region 
2) middle of one region , past a second region , into the middle of a third region
3) beginning of one region into the middle of a second region
4) begining of one region to the end of a second region
5) beginning of one region, through a second region, to the end of a third region
...
etc.

**Review by ADKaster — COMMENTED** [review]
I could be overthinking this, but I definitely think you'll want to write some tests to verify the behavior the manpage is describing.

I think the munmap code is mostly concerned with creating holes. But the mprotect code is concerned with changing the properties of regions in-place, so I'm not sure all the same assumptions hold as the munmap loop over regions.

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:367]: With the missing region and a non existing region I have checekd with [this](https://pastebin.com/FgU3ajMm) and bith state "Cannot allocate Memmory" the code is -1

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:377]: mprotect does not merge (at this point Idk how other systems handle it and It probably is difficult to check)  
and when it is allowed to merge there should be a logicm that it only allows some types of merges  
and munmap carves out what it is supposed to unmap, so the rest of a mapping stays intact

PS:
munmap(ptr, size, prot, flag, fd, off)

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:384]: Just reading the code again, There is an Assertion error on a full match while carving everything out

Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:388]: I've fixed that and may write tests later

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp]: I check the match size now and if it does not add up return EINVAL

**Comment by Hendiadyoin1** [comment]
Gonna hve to rebase to make it work with the red and black tree that is now in place

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp]: There is now a test for it all (no merging still)

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:391]: There is now a test implemented

**Comment by Hendiadyoin1** [comment]
Everything should be working now

File: Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp
Comment (ADKaster) [inline Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp:38]: This file will need moved to Tests/Kernel, after Brian's changes earlier this week.

File: Kernel/Syscalls/mmap.cpp
Comment (ADKaster) [inline Kernel/Syscalls/mmap.cpp:374]: I would expect to see a check similar to:

```c++
        // first check before doing anything that all the intersecting regions are mapped from the same vmobject
        const VMObject* vm_obj = nullptr;
        for (const auto* region : regions) {
            if (!vm_obj) {
                vm_obj = &region->vmobject();
            } else if (&region->vmobject() != vm_obj) {
                return EINVAL;
            }
```
to make sure that all the regions we're trying to potentially merge or change perms of were all from the same vmobject originally.

File: Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp
Comment (ADKaster) [inline Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp:48]: This isn't quite the test I was thinking of. I would expect this test with the 3 discrete mmap call to fail.  If the 6 pages were originally from the same 6-page MAP_ANONYMOUS and mprotect calls had split up the regions in the kernel, then I would expect the mprotects below to pass. (once my other comment is in).

**Review by ADKaster — CHANGES_REQUESTED** [review]
Sorry for letting this almost go stale! Few things that I think should happen before this goes in.

File: Kernel/Syscalls/mmap.cpp
Comment (Hendiadyoin1) [reply Kernel/Syscalls/mmap.cpp:374]: I just looked at the [manpage](https://man7.org/linux/man-pages/man2/mprotect.2.html) again, and it doesn't really seem to specifiy, that you aren't allowed to do that, and due to us not merging anything we don't have to really care, can you look up if we have to have the same vmobject across all regions?

Sidenote: We are returning the wrong code on one or more not mapped pages (EINVAL->ENOMEM)

File: Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp
Comment (Hendiadyoin1) [reply Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp:48]: I've tried it on Linux and the test passes (after some compatability changes and a sudo (for some reason I wasn't allowed to mmap)) so it seems we can hav multiple vm objects or Linux does some optimizations in the background and unifies the vm objects silently...

File: Kernel/Syscalls/mmap.cpp
Comment (ADKaster) [reply Kernel/Syscalls/mmap.cpp:374]: All right, I think it's probably up to the OS then. Doing what Linux does until there's a good reason not to seems fine to me.

File: Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp
Comment (Hendiadyoin1) [reply Userland/Tests/Kernel/mprotect-multi-region-mrotect.cpp]: just rebased and moved it to the corect spot (fixed that return too)

**Comment by Hendiadyoin1** [comment]
If theres nothing else feel free to ping a someone with write access

**Review by ADKaster — APPROVED** [review]

**Comment by Hendiadyoin1** [comment]
@awesomekling @linusg This is ready for a final review now as it seems

**Pull request merged**
